fifo.c

```diff
@@ -2,82 +2,317 @@
+double getdetlatimeofday(struct timeval *, struct timeval *);
+double time_diff(struct timeval , struct timeval );
+
+/* get the delta between two time snapshots */ 
 double
 getdetlatimeofday(struct timeval *begin, struct timeval *end)
 {
     return (end->tv_sec + end->tv_usec * 1.0 / 1000000) -
            (begin->tv_sec + begin->tv_usec * 1.0 / 1000000);
 }
 
+/* determine the latency between two timestamps */
+double 
+time_diff(struct timeval x , struct timeval y)
+{
+    double x_ms , y_ms , diff;
+
+    x_ms = (double)x.tv_sec*1000000 + (double)x.tv_usec;
+    y_ms = (double)y.tv_sec*1000000 + (double)y.tv_usec;
+
+    diff = (double)y_ms - (double)x_ms;
+    return diff;
+}
+
+/* sort the array of doubles (latencies) */
+int compare_double( const void* a, const void* b )
+{
+    if( *(double*)a == *(double*)b ) return 0;
+    return *(double*)a < *(double*)b ? -1 : 1;
+}
+
+/* create a buffer of random upper case letters */
+int populate_buf(char *buffer, int length) {
+
+    int i = length - 1;
+    int r = rand() % 26;
+
+    buffer[i] = '\0';
+
+    while ( SIZE_OF_USED_BUFFER != i) {
+        i = i - 1;
+        buffer[i] = 'A' + r;
+        r++;
+        if (r==26)
+           r = 0;
+    }
+    return 0;
+}
+
+/* ================================================================================================================================= */
+
 int main(int argc, char *argv[]) {
-    int              fd;
-    int              i, size, count, sum, n;
-    char            *buf;
+    int              fd1, fd2;
+    long             msgtot, size, sum, n;
+    short 	     mins;
+    long 	     ctr;
+    short            runnum;
+
+    struct mybuf {
+	    int seq;
+	    struct timeval tv_time_sent;
+	    char *rest;
+    } *buf;
+
+    time_t start;
+
+    char latencies_file[50];
+    FILE *outfile_latencies;
+
     struct timeval   begin, end;
+    struct timeval  tv_post_read;
+
+    double time_elapsed;
+    double avg_elapsed, tot_elapsed;
 
-    if (argc != 3) {
-        printf("usage: ./fifo <size> <count>\n");
+    int ninetyninth, ninetyfifth;  // 95th% and 99th% latencies
+
+    if (argc != 4) {
+        printf("usage: ./fifo <message-size> <mins-to-run> <run-number>\n");
         return 1;
     }
 
+    typedef struct  {
+            double start_time;
+            double end_time;
+    } latencies_arr;
+
     size = atoi(argv[1]);
-    count = atoi(argv[2]);
+    mins = atoi(argv[2]);
+    runnum = atoi(argv[3]);
+
+    int new_size = ARRAY_BOOST;
+    
+    double *latencies = (double*) malloc(ARRAY_BOOST * sizeof(double));
+    if (latencies == NULL) {
+        printf("Latencies memory not allocated.\n");
+        return 1;
+    }
+    latencies_arr *lat = (latencies_arr*) malloc(ARRAY_BOOST * sizeof(latencies_arr));
+    if (lat == NULL) {
+        printf("Latencies array memory not allocated.\n");
+        return 1;
+    }
+
+    snprintf(latencies_file,50,"fifo_latencies_%d_%hi.json", (int) size, runnum);
```

Comment:
I'm not sure I understand the purpose of the `runnum` value. We are just taking it as input and returning it back in the output?